ui/src/BaseEditor.ts

```diff
@@ -0,0 +1,166 @@
+import { localize } from "./localize";
+import { TemporaryFile } from './utils/TemporaryFile';
+
+// tslint:disable-next-line:typedef
+const dialogResponses = {
+    OK: localize('OK', "OK"),
+    DontShowAgain: localize('dontShow', "Don't Show Again")
+};
+
+// tslint:disable-next-line:no-unsafe-any
+export abstract class BaseEditor<ContextT> implements vscode.Disposable {
+    private fileMap: { [key: string]: [vscode.TextDocument, ContextT] } = {};
+    private ignoreSave: boolean = false;
+
+    constructor(readonly showSavePromptKey: string, readonly outputChanel?: vscode.OutputChannel) {
+    }
+
+    public abstract getData(context: ContextT): Promise<string>;
+    public abstract updateData(context: ContextT, data: string): Promise<string>;
+    public abstract getFilename(context: ContextT): Promise<string>;
+    public abstract getSize(context: ContextT): Promise<number>;
+    public abstract getSaveConfirmationText(context: ContextT): Promise<string>;
+
+    public async showEditor(context: ContextT): Promise<void> {
+        const size: number = await this.getSize(context);
+        const fileName: string = await this.getFilename(context);
+        const splitFileName: string[] = fileName.split('.');
+        const extension: string = splitFileName[splitFileName.length - 1];
+
+        this.appendToOutput(localize('opening', 'Opening "{0}" ...', fileName));
+        if (size > 50 /*Megabytes*/) {
+            const message: string = localize('tooLargeError', '"{0}" is too large to download.', fileName);
+
+            await vscode.window.showWarningMessage(message, dialogResponses.OK);
+            this.appendLineToOutput(localize('failed', "Failed."));
+            this.appendLineToOutput(localize('errorDetails', 'Error Details: {0}', message));
+        } else if (extension === 'exe' || extension === 'img' || extension === 'zip') {
+            const message: string = localize('unsupportedError', '"{0}" has an unsupported file extension.', fileName);
+
+            await vscode.window.showWarningMessage(message, dialogResponses.OK);
+            this.appendLineToOutput(localize('failed', " Failed."));
+            this.appendLineToOutput(localize('errorDetails', 'Error Details: {0}', message));
+        } else {
+            try {
+                // tslint:disable-next-line:no-unsafe-any
+                const localFilePath: string = await TemporaryFile.create(fileName);
+                const document: vscode.TextDocument = await vscode.workspace.openTextDocument(localFilePath);
+                this.fileMap[localFilePath] = [document, context];
+                const data: string = await this.getData(context);
+                const textEditor: vscode.TextEditor = await vscode.window.showTextDocument(document);
+                await this.updateEditor(data, textEditor);
+                this.appendLineToOutput(' Done.');
+            } catch (error) {
+                let details: string;

```

Comment:
Again, I would get rid of all the error handling in this file and let each extension handle it